pkg/k8srbac/subjects_test.go

```diff
@@ -0,0 +1,106 @@
+	"github.com/stackrox/rox/pkg/protoassert"
+)
+
+func TestGetSubjectsAdjustedByKind(t *testing.T) {
+	tests := map[string]struct {
+		rb     *storage.K8SRoleBinding
+		expect []*storage.Subject
+	}{
+		"nil rolebinding": {
+			rb:     nil,
+			expect: nil,
+		},
+		"nil subjects": {
+			rb:     &storage.K8SRoleBinding{},
+			expect: nil,
+		},
+		"empty subjects": {
+			rb:     &storage.K8SRoleBinding{Subjects: make([]*storage.Subject, 0)},
+			expect: nil,
+		},
+		"all kinds supported": {
+			rb: &storage.K8SRoleBinding{Subjects: []*storage.Subject{
+				{Kind: storage.SubjectKind_UNSET_KIND},
+				{Kind: storage.SubjectKind_SERVICE_ACCOUNT},
+				{Kind: storage.SubjectKind_USER},
+				{Kind: storage.SubjectKind_GROUP},
+			}},
+			expect: []*storage.Subject{
+				{Kind: storage.SubjectKind_UNSET_KIND},
+				{Kind: storage.SubjectKind_SERVICE_ACCOUNT},
+				{Kind: storage.SubjectKind_USER},
+				{Kind: storage.SubjectKind_GROUP},
+			},
+		},
+		"namespaced kinds preserve namespace": {
+			rb: &storage.K8SRoleBinding{Subjects: []*storage.Subject{
+				{Kind: storage.SubjectKind_UNSET_KIND, Namespace: "namespace"},
+				{Kind: storage.SubjectKind_SERVICE_ACCOUNT, Namespace: "namespace"},
+			}},
+			expect: []*storage.Subject{
+				{Kind: storage.SubjectKind_UNSET_KIND, Namespace: "namespace"},
+				{Kind: storage.SubjectKind_SERVICE_ACCOUNT, Namespace: "namespace"},
+			},
+		},
+		"non-namespaced kinds are adjusted": {
+			rb: &storage.K8SRoleBinding{Subjects: []*storage.Subject{
+				{Kind: storage.SubjectKind_USER, Namespace: "namespace"},
+				{Kind: storage.SubjectKind_GROUP, Namespace: "namespace"},
+			}},
+			expect: []*storage.Subject{
+				{Kind: storage.SubjectKind_USER},
+				{Kind: storage.SubjectKind_GROUP},
+			},
+		},
+		"only non-namespaced kinds are adjusted for list of mixed kinds": {
+			rb: &storage.K8SRoleBinding{Subjects: []*storage.Subject{
+				{Kind: storage.SubjectKind_UNSET_KIND, Namespace: "namespace"},
+				{Kind: storage.SubjectKind_UNSET_KIND},
+				{Kind: storage.SubjectKind_SERVICE_ACCOUNT, Namespace: "namespace"},
+				{Kind: storage.SubjectKind_USER, Namespace: "namespace"},
+				{Kind: storage.SubjectKind_GROUP, Namespace: "namespace"},
+			}},
+			expect: []*storage.Subject{
+				{Kind: storage.SubjectKind_UNSET_KIND, Namespace: "namespace"},
+				{Kind: storage.SubjectKind_UNSET_KIND},
+				{Kind: storage.SubjectKind_SERVICE_ACCOUNT, Namespace: "namespace"},
+				{Kind: storage.SubjectKind_USER},
+				{Kind: storage.SubjectKind_GROUP},
+			},
+		},
+		"non-namespaced duplicates are adjusted": {
+			rb: &storage.K8SRoleBinding{Subjects: []*storage.Subject{
+				{Kind: storage.SubjectKind_USER, Namespace: "namespace"},
+				{Kind: storage.SubjectKind_USER},
+				{Kind: storage.SubjectKind_GROUP, Namespace: "namespace"},
+				{Kind: storage.SubjectKind_GROUP},
+			}},
+			expect: []*storage.Subject{
+				{Kind: storage.SubjectKind_USER},
+				{Kind: storage.SubjectKind_GROUP},
+			},
```

Comment:
mega-nitpick (to show that I reviewed this properly and not just gave ✅ ): we could also make sure that the order does not play a role and maybe do:

```suggestion
		"non-namespaced duplicates are adjusted": {
			rb: &storage.K8SRoleBinding{Subjects: []*storage.Subject{
				{Kind: storage.SubjectKind_USER, Namespace: "namespace"},
				{Kind: storage.SubjectKind_USER},
				{Kind: storage.SubjectKind_GROUP},
				{Kind: storage.SubjectKind_GROUP, Namespace: "namespace"},
			}},
			expect: []*storage.Subject{
				{Kind: storage.SubjectKind_USER},
				{Kind: storage.SubjectKind_GROUP},
			},
```

Reply:
I like that. I made it on purpose that the first record has a namespace to ensure that the first record is not used. But you are right, it would be nice to check both cases.

Reply:
Thanks for improving!